### ISSUE TITLE
Property: can be dropped if is equals to parent declaration

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#333]
It should detect if the property have the same value that the parent declaration.

```php
class ParentClass                    { protected $casts = []; }
class ChildClass extends ParentClass { protected $casts = []; } // Warn.
```

Note that `ChildClass` declares `$casts` with the same value from `ParentClass`, then `$casts` could by dropped without side-effects.

To consider:

* Traits properties should be considered;
* If property visiblity is changed, then it is acceptable;

**Comment by rentalhost** [comment]
At this trail, there are some API that I can found the parent declaration from a method or property? I mean something like that:

```java
public static Method findParentDeclaration(Methodreference);
public static Field findParentDeclaration(Field reference);
```

Then I could uses like:

```java
@Override
public void visitPhpField(Field field) {
    final Field parentDeclaration = findParentDeclaration(field);

    if (field != null) {
        // Compare here.
    }
}
```

It should be very helpful in cases like that.

(ping @kalessil for that comment)

**Comment by kalessil** [comment]
Don't we have an inspection complaining about properties override?

**Comment by rentalhost** [comment]
I don't know. I feel that yes, but I can't found (at least by searching for "property" on inspections).

**Comment by funivan** [comment]
@rentalhost try to search `field` 
https://github.com/kalessil/phpinspectionsea/blob/master/docs/architecture.md#class-overrides-a-field-of-a-parent-class

**Comment by rentalhost** [comment]
Okay, it does de job, except by two things:

**1. It will check if you are overring property:** and the objective is check only if values are the same, or then warn it only if the parent declaration is not falsy (we should ignores boolean here, because it is just a two state value, then both `true`/`false `can be modified to `false`/`true`, respectively).

```php
// Acceptable: parent declaration is falsy.
class ParentClass { public $isFalsy = []; }
class ChildClass extends ParentClass { public $isFalsy = [ 1, 2, 3 ]; }

// Not acceptable: parent declaration is not falsy.
class ParentClass { public $isFalsy = [ 1, 2, 3 ]; }
class ChildClass extends ParentClass { public $isFalsy = [ 1, 2, 3, 4, 5 ]; }

// Not acceptable: parent declaration type is diferent of child declaration.
// Even if it is falsy-to-falsy types, but strictly different.
class ParentClass { public $isFalsy = []; }
class ChildClass extends ParentClass { public $isFalsy = false; }
```

It is useful because when we are working with frameworks like Laravel we should change some properties directly on properties (example: `$table = "my_table"` from `Eloquent` class). Else we should do that on `__construct()`, but it is something that is not "recommended" from [Laravel docs](https://laravel.com/docs/5.4/eloquent#eloquent-model-conventions).

**2. It doesn't consider indirect traits:** at least on current implementation.

```php
trait MyTrait { public $casts = [ 1, 2, 3 ]; }

// Acceptable, but should not.
class ParentClass                    { use MyTrait; }
class ChildClass extends ParentClass { public $casts = [ 1, 2, 3 ]; }
```

---

My suggestion is create an additional option:

**Check for any overring:** to keep current behaviour. It is, if in any case you redeclare a property, then you will be warned. When disabled it will check only for override that modify truth parent declaration (except for boolean) or if you modify the original type (that is not compatible with `@type`).

**Comment by kalessil** [comment]
Traits support makes sense, I think in the existing inspection.
And regarding defaults: this probably should o in field initialization flaws inspection.

And to plan: I'll ping I'd like to do it by myself - I have some notes in my off-line list for the both inspections.

**Comment by kalessil** [comment]
Picking up for 2.3.8 - will be easier to keep the focus on other complex tasks afterward.

**Comment by kalessil** [comment]
Status update:
- traits support in field overrides/inti flaws there
- init flaws now reports duplicate overrides of any type

And regarding the laravel specific: it makes sense to make a FR to the @Haehnchen here: https://github.com/Haehnchen/idea-php-laravel-plugin

PS: @rentalhost, I pushed a fresh binary, you can grab the mentioned chnages from master
